### PR TITLE
Feature/ios javascript interface implementation

### PR DESCRIPTION
**Pull request by ts-leojoseph-r**
# Description
Native iOS and Javascript bridge implementation.
Following changes are done,
* Podspec change, 
* Javascript that is required for bridging
* Script message handler
* Callback response to Mini app

## Links
[MINI-609](https://jira.rakuten-it.com/jira/browse/MINI-609)

# Checklist
- [x] I have read the [contributing guidelines](../CONTRIBUTING.md).
- [x] I wrote/updated tests for new/changed code
- [x] I ran `fastlane ci` without errors

### CODE REVIEW

File: MiniApp/Classes/Display/RealMiniAppView.swift [hunk 1]
```diff
@@ -1,15 +1,18 @@
 import WebKit
 
-public class RealMiniAppView: UIView, MiniAppDisplayProtocol {
-
-    private var webView: WKWebView
+public class RealMiniAppView: UIView, MiniAppDisplayProtocol, MiniAppCallbackProtocol {
+    internal var webView: WKWebView
+    private var messageInterface: MiniAppMessageProtocol
```
Comment (Climbatize) [inline MiniApp/Classes/Display/RealMiniAppView.swift:5]: shouldn't we use delegate instead of interface ?

Comment (ts-leojoseph-r) [reply MiniApp/Classes/Display/RealMiniAppView.swift:5]: I guess you are right, I need to change this and also I think we need to have a different name altogether which confuses us with Mini App (Web App) interface name as well.

Comment (ts-leojoseph-r) [reply MiniApp/Classes/Display/RealMiniAppView.swift]: Done

File: MiniApp/Classes/Display/RealMiniAppView.swift [hunk 2]
```diff
@@ -25,4 +28,16 @@ public class RealMiniAppView: UIView, MiniAppDisplayProtocol {
     public func getMiniAppView() -> UIView {
         return self
     }
+
+    func didRecieveScriptMessageResponse(messageId: String, response: String) {
```
Comment (donnie-jp) [inline MiniApp/Classes/Display/RealMiniAppView.swift:32]: Recieve -> Receive

File: MiniApp/Classes/Extensions/MiniApp+WKUserContentController.swift [hunk 3]
```diff
@@ -0,0 +1,29 @@
...
+    /// Method to add Mini App custom WKScriptMessageHandler.
+    /// - Parameters:
+    ///   - delegate: Callback protocol to handle the message from the WebView
+    ///   - hostMessageInterface: Message interface protocol of the host application
+    func addMiniAppScriptMessageHandler(delegate: MiniAppCallbackProtocol, hostAppMessageDelegate: MiniAppMessageProtocol) {
+        add(MiniAppScriptMessageHandler(delegate: delegate, hostAppMessageDelegate: hostAppMessageDelegate), name: Constants.javascriptInterfaceName)
+    }
+
+    /// Method to remove the custom WKScriptMessageHandler. Removing the message handler will help to avoid memory leaks
+    func removeMessageHandler() {
```
Comment (donnie-jp) [inline MiniApp/Classes/Extensions/MiniApp+WKUserContentController.swift:14]: Can the Controller's deinit be extended to automatically remove the messaged handler?

File: MiniApp/Classes/Extensions/MiniApp+WKUserContentController.swift [hunk 4]
```diff
@@ -0,0 +1,29 @@
...
+        removeScriptMessageHandler(forName: Constants.javascriptInterfaceName)
+    }
+
+    /// Method to add the Briding Javascript to the WebView usercontroller
+    /// This JavaScript is responsible for communication between SDK and Mini App
+    /// Any request from Mini App is communcated via the contract available in the Bridging script
```
Comment (donnie-jp) [inline MiniApp/Classes/Extensions/MiniApp+WKUserContentController.swift:20]: communcated -> communicated

File: MiniApp/Classes/Extensions/MiniApp+WKUserContentController.swift [hunk 5]
```diff
@@ -0,0 +1,29 @@
...
+    func addBridgingJavaScript() {
+        let podBundle = Bundle(for: MiniApp.self)
```
Comment (donnie-jp) [inline MiniApp/Classes/Extensions/MiniApp+WKUserContentController.swift:22]: To test this properly you probably need a `Bundle` (or equivalent protocol) param passed in with a default of `MiniApp.self`.

File: MiniApp/Classes/JavascriptBridge/MiniAppScriptMessageHandler.swift [hunk 6]
```diff
@@ -0,0 +1,60 @@
...
+
+protocol MiniAppCallbackProtocol: AnyObject {
+    func didRecieveScriptMessageResponse(messageId: String, response: String)
+    func didRecieveScriptMessageError(messageId: String, errorMessage: String)
+}
+
+internal class MiniAppScriptMessageHandler: NSObject, WKScriptMessageHandler {
+
+    weak var delegate: MiniAppCallbackProtocol?
+    var hostAppMessageDelegate: MiniAppMessageProtocol?
```
Comment (donnie-jp) [inline MiniApp/Classes/JavascriptBridge/MiniAppScriptMessageHandler.swift:12]: Does this delegate need to be weak?

File: MiniApp/Classes/JavascriptBridge/MiniAppScriptMessageHandler.swift [hunk 7]
```diff
@@ -0,0 +1,60 @@
...
+
+    init(delegate: MiniAppCallbackProtocol, hostAppMessageDelegate: MiniAppMessageProtocol) {
+        self.delegate = delegate
+        self.hostAppMessageDelegate = hostAppMessageDelegate
+        super.init()
+    }
+
+    public func userContentController(_ userContentController: WKUserContentController, didReceive message: WKScriptMessage) {
```
Comment (donnie-jp) [inline MiniApp/Classes/JavascriptBridge/MiniAppScriptMessageHandler.swift:20]: Can this public API go in its own extension?

File: MiniApp/Classes/JavascriptBridge/MiniAppScriptMessageHandler.swift (hunk 6 above)
Comment (Climbatize) [reply MiniApp/Classes/JavascriptBridge/MiniAppScriptMessageHandler.swift:12]: yes it preferably does, this avoids reference cycles

File: MiniApp/Classes/Extensions/MiniApp+WKUserContentController.swift (hunk 3 above)
Comment (Climbatize) [reply MiniApp/Classes/Extensions/MiniApp+WKUserContentController.swift:14]: @donnie-jp it's a good idea. But it's not really possible here. This expected answer on S.O has a workaround to make the messageHandler being deallocated.

File: MiniApp/Classes/Display/RealMiniAppView.swift (hunk 2 above)
Comment (ts-leojoseph-r) [reply MiniApp/Classes/Display/RealMiniAppView.swift]: 🙈Done

File: MiniApp/Classes/Extensions/MiniApp+WKUserContentController.swift (hunk 4 above)
Comment (ts-leojoseph-r) [reply MiniApp/Classes/Extensions/MiniApp+WKUserContentController.swift]: 🙈Done

File: MiniApp/Classes/Extensions/MiniApp+WKUserContentController.swift (hunk 3 above)
Comment (donnie-jp) [reply MiniApp/Classes/Extensions/MiniApp+WKUserContentController.swift:14]: Thanks. Ok lets not introduce a hacky workaround and keep the explicit remove 🙂

File: MiniApp/Classes/Extensions/MiniApp+WKUserContentController.swift (hunk 5 above)
Comment (ts-leojoseph-r) [reply MiniApp/Classes/Extensions/MiniApp+WKUserContentController.swift]: Done 👍

File: MiniApp/Classes/JavascriptBridge/MiniAppScriptMessageHandler.swift (hunk 6 above)
Comment (ts-leojoseph-r) [reply MiniApp/Classes/JavascriptBridge/MiniAppScriptMessageHandler.swift]: Done 👍

File: MiniApp/Classes/Extensions/MiniApp+WKUserContentController.swift (hunk 3 above)
Comment (ts-leojoseph-r) [reply MiniApp/Classes/Extensions/MiniApp+WKUserContentController.swift:14]: 👍

File: MiniApp/Classes/JavascriptBridge/MiniAppScriptMessageHandler.swift (hunk 7 above)
Comment (ts-leojoseph-r) [reply MiniApp/Classes/JavascriptBridge/MiniAppScriptMessageHandler.swift]: No, we can't. While adding a script handler to the user controller, we need a class that conforms to WKScriptMessageHandler. That's why I created a separate class MiniAppScriptMessageHandler

**Pull request merged**